### PR TITLE
feat(media-query): add additional hooks

### PR DESCRIPTION
**Pull request by with-heart**
This PR adds the following `media-query` hooks:

- `useBreakpointValue`
- `useAnimationPreference`
- `useColorModePreference`
- `useViewportSize`

### CODE REVIEW

File: packages/media-query/src/media-query.hook.tsx [hunk 1]
```diff
@@ -148,3 +150,82 @@ const createMediaQueries = (breakpoints: Dict) => {
     return mediaQuery
   })
 }
+
+type BreakpointValues<T> = { [breakpoint: string]: T }
```
Comment (with-heart) [inline packages/media-query/src/media-query.hook.tsx:154]: Should we include the responsive array style as well?

File: packages/media-query/src/media-query.hook.tsx [hunk 2]
```diff
@@ -148,3 +150,82 @@ const createMediaQueries = (breakpoints: Dict) => {
...
+  const breakpoint = useBreakpoint()
+
+  // if breakpoint is undefined, we'll return undefined
+  if (!breakpoint) return
+
+  // if we have no value for the current breakpoint, return the `base`
+  // breakpoint value
+  const baseValue = values.base
+
+  return values[breakpoint] || baseValue
```
Comment (with-heart) [inline packages/media-query/src/media-query.hook.tsx:175]: Do we already have tooling somewhere for finding a value from a breakpoints object that defaults to `base`? I saw `assignObject` but that wasn't quite what we needed here.

File: packages/media-query/src/media-query.hook.tsx [hunk 3]
```diff
@@ -148,3 +150,82 @@ const createMediaQueries = (breakpoints: Dict) => {
...
+export function useViewportSize(
+  ref: React.RefObject<HTMLElement>,
+): Readonly<[number, number]> {
+  const [size, setSize] = useState<Readonly<[number, number]>>([0, 0])
+
+  useEffect(() => {
+    if (!ref.current) return undefined
+
+    const observer = new ResizeObserver(([entry]) => {
+      const { inlineSize: width, blockSize: height } = entry.contentBoxSize[0]
```
Comment (with-heart) [inline packages/media-query/src/media-query.hook.tsx:220]: `ResizeObserver` gives us `contentBoxSize` or `borderBoxSize`, and I'm not sure which would be more appropriate here. Should this include the padding/border? Or maybe we should return the width/height for both and let the user decide?

Comment (segunadebayo) [reply packages/media-query/src/media-query.hook.tsx:220]: Should we consider using "@reach/rect" for this part? It's much smaller than the resize observer package and works really well.

What do you think? I feel we can simply re-export `useRect` from `@reach/rect`

File: packages/media-query/src/media-query.hook.tsx (hunk 1 above)
Comment (segunadebayo) [reply packages/media-query/src/media-query.hook.tsx:154]: We're enforcing breakpoints in the theme to be an object in v1. It makes the logic a lot easier.

File: packages/media-query/src/media-query.hook.tsx (hunk 2 above)
Comment (segunadebayo) [reply packages/media-query/src/media-query.hook.tsx:175]: Oh no. We need to make a decision on this. I was thinking on calling it either `base` or `xs`. Then we can update the docs and code to be consistent

Comment (with-heart) [reply packages/media-query/src/media-query.hook.tsx:175]: Is it just trying to figure out the right now or would they work differently?

File: packages/media-query/src/media-query.hook.tsx (hunk 3 above)
Comment (with-heart) [reply packages/media-query/src/media-query.hook.tsx:220]: That's a better idea than using the larger package. Though maybe the best idea is not even re-exporting and letting users decide which to use if they want this functionality?

File: packages/media-query/src/media-query.hook.tsx (hunk 2 above)
Comment (segunadebayo) [reply packages/media-query/src/media-query.hook.tsx:175]: Let's leave it at `base` for now.

**Comment by segunadebayo** [comment]
Again, please rebase this with `master` and send a PR

**Comment by PaulPCIO** [comment]
image

This happens when resizing the window?

**Comment by ljosberinn** [comment]
@PaulPCIO hi, is this related to this PR or is it a separate issue?

**Comment by PaulPCIO** [comment]
Hi, no sorry it's unrelated, I was about to open up an issue but this seems like a ghost in the machine, I tried replicating this again but it does not seem to happen anymore.  ¯\_(ツ)_/¯

**Pull request closed without merging**